src/chainstate/burn/db/sortdb.rs

```diff
@@ -992,6 +1017,92 @@ impl <'a> SortitionHandleConn <'a> {
+        assert!(pox_consts.anchor_threshold > (pox_consts.prepare_length / 2));
+
+        // if this block is the _end_ of a prepare phase,
+        if ! (my_height % pox_consts.reward_cycle_length == 0) {
+            return Err(CoordinatorError::NotPrepareEndBlock)
+        }
+
+        let prepare_end = my_height;
+        // if this block isn't greater than prepare_length, then this shouldn't be the end of a prepare block
+        if prepare_end < pox_consts.prepare_length {
+            return Err(CoordinatorError::NotPrepareEndBlock)
+        }
+
+        let prepare_begin = prepare_end - pox_consts.prepare_length;
+
+        let mut candidate_anchors = HashMap::new();
+        let mut memoized_candidates: HashMap<_, (Txid, u64)> = HashMap::new();
+
+        // iterate over every sortition winner in the prepare phase
+        //   looking for their highest ancestor _before_ prepare_begin.
+        let winners = self.get_sortition_winners_in_fork(prepare_begin, prepare_end)?;
+        for (winner_commit_txid, winner_block_height) in winners.into_iter() {
+            let mut cursor = (winner_commit_txid, winner_block_height);
+
+            while cursor.1 > (prepare_begin as u64) {
+                // check if we've already discovered the candidate for this block
+                if let Some(ancestor) = memoized_candidates.get(&cursor.1) {
+                    cursor = ancestor.clone();
+                } else {
+                    // get the block commit
+                    let block_commit = self.get_block_commit_by_txid(&cursor.0)?
+                        .expect("CORRUPTED: Failed to fetch block commit for known sortition winner");
+                    // find the parent sortition
+                    let sn = SortitionDB::get_ancestor_snapshot(self, block_commit.parent_block_ptr as u64, &self.context.chain_tip)?
+                        .expect("CORRUPTED: accepted block commit, but parent pointer not in sortition set");
+                    assert!(sn.sortition, "CORRUPTED: accepted block commit, but parent pointer not a sortition winner");
+
+                    cursor = (sn.winning_block_txid, sn.block_height);
+                }
+            }
+            // this is the burn block height of the sortition that chose the
+            //   highest ancestor of winner_stacks_bh whose sortition occurred before prepare_begin
+            //  the winner of that sortition is the PoX anchor block candidate that winner_stacks_bh is "voting for"
+            let highest_ancestor = cursor.1;
+            memoized_candidates.insert(winner_block_height, cursor);
+            if let Some(x) = candidate_anchors.get_mut(&highest_ancestor) {
+                *x += 1;
+            } else {
+                candidate_anchors.insert(highest_ancestor, 1u32);
+            }
+        }
+
+        // did any candidate receive >= F*w?
+        for (candidate, confirmed_by) in candidate_anchors.into_iter() {
```

Comment:
If it's not too much trouble, can you also add a runtime assert here that there is at most one block for which `confirmed_by >= pox_constants.anchor_threshold`?